_posts/2020-10-14-threshold-tuning/2020-10-14-threshold-tuning.Rmd

```diff
@@ -0,0 +1,231 @@
+  assigning an observation to a certain class. This can lead to improved ...
+output:
+  distill::distill_article:
+    self_contained: false
+    css: ../../custom.css
+---
+
+```{r, include = FALSE}
+## Just some preparation
+knitr::opts_chunk$set(
+  echo = TRUE,
+  R.options = list(width = 80)
+)
+data.table::setDTthreads(1)
+set.seed(20191101)
+lgr::get_logger("mlr3")$set_threshold("warn")
+```
+
+## Intro
+
+Predicting probabilities in classification tasks allows us to adjust the probability thresholds required for assigning an observation to a certain class.
+This can lead to improved classification performance, especially for cases where we e.g. aim to balance off metrics such as false positive and false negative rates.
+
+This is for example often done in ROC Analysis. The mlr3book also has a chapter on [ROC Analysis](https://mlr3book.mlr-org.com/binary-classification.html#binary-roc)) for the interested reader. This post does not focus on ROC analysis, but instead focusses on the general problem of adjusting classification thresholds for arbitrary metrics.
+
+
+This post assumes some familiarity with the `mlr3`, `mlr3pipelines` and `mlr3tuning` package, as both are used during the post.
+The [mlr3book](https://mlr3book.mlr-org.com/) contains more details on those two packages.
+This post is a more in-depth version of the [article on threshold tuning in the mlr3book](https://mlr3book.mlr-org.com/cost-sens.html#threshold-tuning-1).
+
+
+## Prerequisites
+
+Before we start, we have load all required packages:
+
+```{r packages}
+library(mlr3)
+library(mlr3pipelines)
+library(mlr3tuning)
+```
+
+
+
+## Thresholds: A short intro
+
+In order to understand thresholds, we will quickly showcase the effect of setting different thresholds:
+
+First we create a learner that predicts probabilities and use it to predict on holdout data, storing the prediction.
+
+```{r}
+learner = lrn("classif.rpart", predict_type = "prob")
+rr = resample(tsk('pima'), learner, rsmp("holdout"))
+prd = rr$prediction()
+prd
+```
+
+If we now look at the confusion matrix, the off-diagonal elements are errors made by our model (*false positives* and *true negatives*) while on-diago)ements are where our model predicted correctly.
+
+
+```{r}
+# Print confusion matrix
+prd$confusion
+# Print False Positives and False Negatives
+prd$score(list(msr("classif.fp"), msr("classif.fn")))
+```
+
+By adjusting the **classification threshold**, in this case the probability required to predict the positive class, we can now trade off predicting more positive cases (first row)
+against predicting fewer negative cases (second row) or vice versa.
+
+```{r}
+# Lower threshold: More positives
+prd$set_threshold(0.25)$confusion
+```
+
+```{r}
+# Higher threshold: Fewer positives
+prd$set_threshold(0.75)$confusion
+```
+
+This threshold value can now be adjusted optimially for a given measure, such as accuracy. How this can be done is discussed in the following section.
```

Comment:
```suggestion
This threshold value can now be adjusted optimally for a given measure, such as accuracy. How this can be done is discussed in the following section.
```